src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutorKeepalived.java

```diff
@@ -0,0 +1,308 @@
+        // which will then call retrier again. It will reset the retry time counter so we could
+        // retry more than --remote_retry times which is not expected.
+        response = retrier
+            .execute(() -> Utils.refreshIfUnauthenticated(this::execute, callCredentialsProvider),
+                backoff);
+
+        // If no response from Execute(), use WaitExecution() in a "loop" which is implicated inside
+        // the retry block.
+        if (response == null) {
+          response = retrier.execute(
+              () -> Utils.refreshIfUnauthenticated(this::waitExecution, callCredentialsProvider),
+              backoff);
+        }
+      }
+
+      return response;
+    }
+
+    @Nullable
+    ExecuteResponse execute() throws IOException {
+      Preconditions.checkState(lastOperation == null);
+
+      try {
+        Iterator<Operation> streamOperations = executionBlockingStubSupplier.get().execute(request);
+
+        // We don't want to reset backoff for Execute() since if there is an error:
+        //   1. If happened before we received a first response, we want to ensure the retry counter
+        //      is increased and call Execute() again.
+        //   2. Otherwise, we will fallback to WaitExecution() loop.
+        return handleStreamOperations(streamOperations, /* resetBackoff */ false);
+      } catch (StatusRuntimeException e) {
+        if (lastOperation != null) {
+          // By returning null, we are going to fallback to WaitExecution() loop.
+          return null;
+        }
+        throw new IOException(e);
+      }
+    }
+
+    @Nullable
+    ExecuteResponse waitExecution() throws IOException {
+      Preconditions.checkState(lastOperation != null);
+
+      WaitExecutionRequest request = WaitExecutionRequest.newBuilder()
+          .setName(lastOperation.getName())
+          .build();
+      try {
+        Iterator<Operation> streamOperations = executionBlockingStubSupplier.get()
+            .waitExecution(request);
+        // We want to reset backoff for WaitExecution() so we can "infinitely" wait for the
+        // execution to complete as long as they are making progress (by returning response).
+        return handleStreamOperations(streamOperations, /* resetBackoff */ true);
+      } catch (StatusRuntimeException e) {
+        if (e.getStatus().getCode() == Code.NOT_FOUND) {
+          // Operation was lost on the server. Retry Execute.
+          lastOperation = null;
+          return null;
+        }
+        throw new IOException(e);
+      }
+    }
+
+    /**
+     * Process a stream of operations from Execute() or WaitExecution().
+     *
+     * @param resetBackoff reset backoff if true once received a response successfully.
+     */
+    ExecuteResponse handleStreamOperations(Iterator<Operation> streamOperations,
+        boolean resetBackoff) throws IOException {
+      try {
+        while (streamOperations.hasNext()) {
```

Comment:
What is the difference between getting a new `Operation` on the stream returned by `execute()` and the stream finishing and requiring us to call `waitExecution()` again?

My theory: this is so that we can have actions that run longer than the maximum allowed time for an RPC but then why stream `Operation` messages at all instead of just returning one and requiring to call `waitExecution()` in a loop?

Reply:
According to the [specification](https://source.corp.google.com/piper///depot/google3/third_party/bazel_remoteapis/build/bazel/remote/execution/v2/remote_execution.proto), server can return multiple `Operation`s using the stream but is not required. After the first response, the stream we get from Execute() is the same as what we can get from WaitExecution().

If server implementation send `Operation`s over the stream periodically, we don't need `WaitExecution` in normal cases as you said.

However, stream can finish due to network issues or simply caused by timeout. In this case, we have to use WaitExecution().

So why Execute() return a stream of `Operation` instead of one? I am not sure but this is defined by REAPI.

Reply:
If the spec says so, the spec says so... it sure looks weird and I was hoping you could tell me why this strange pattern exists for my education. From what I see, it's not at all common anywhere else (but exists in a few places). In any case, this is not relevant for this code review.